### PR TITLE
Clear `eval_libc` errors from unix shims

### PR DESCRIPTION
**Pull request by noahmbright**
Each instance of `eval_libc("E` is scrubbed out. Currently, this has the idealistic function calls, but this won't compile yet. Marking as a draft and making comments with the representative compilation errors. Whether we want to hold off on making these erring changes or address the root problem now is up to the reviewers. 

See comments for the issues concretely, but it seems like everything stems from differences between the return types of `this.eval_libc` and `LibcError`.

### CODE REVIEW

File: src/shims/alloc.rs [hunk 1]
```diff
@@ -109,7 +109,7 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
         // Align must be power of 2, and also at least ptr-sized (POSIX rules).
         // But failure to adhere to this is not UB, it's an error condition.
         if !align.is_power_of_two() || align < this.pointer_size().bytes() {
-            interp_ok(this.eval_libc("EINVAL"))
+            interp_ok(LibcError("EINVAL"))
```
Comment (noahmbright) [inline src/shims/alloc.rs:112]: Passing a `LibcError("Efoo")` to `interp_ok` results in 
```expected Scalar<Provenance>, found IoError```

File: src/shims/unix/foreign_items.rs [hunk 2]
```diff
@@ -653,13 +652,10 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
                 let chunk_size = CpuAffinityMask::chunk_size(this);
 
                 if this.ptr_is_null(mask)? {
-                    let efault = this.eval_libc("EFAULT");
-                    this.set_last_error(efault)?;
-                    this.write_int(-1, dest)?;
+                    this.set_last_error_and_return(LibcError("EFAULT"), dest);
```
Comment (noahmbright) [inline src/shims/unix/foreign_items.rs:655]: I think semantically `set_last_error_and_return` is the reduction of this pattern, but please double check.

File: src/shims/unix/foreign_items.rs [hunk 3]
```diff
@@ -936,7 +925,7 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
                     this.write_null(dest)?;
                 } else {
                     this.write_null(&result)?;
-                    this.write_scalar(this.eval_libc("ERANGE"), dest)?;
+                    this.write_scalar(LibcError("ERANGE"), dest)?;
```
Comment (noahmbright) [inline src/shims/unix/foreign_items.rs:928]: The issue here is that `From<IoError>` is not implemented for `rustc_const_eval::interpret::Scalar<machine::Provenance>`. This can probably be handled in another commit.

File: src/shims/unix/linux/foreign_items.rs [hunk 4]
```diff
@@ -85,7 +85,7 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
                     this.read_scalar(name)?,
                     TASK_COMM_LEN,
                 )?;
-                let res = if res { Scalar::from_u32(0) } else { this.eval_libc("ERANGE") };
+                let res = if res { Scalar::from_u32(0) } else { LibcError("ERANGE") };
```
Comment (noahmbright) [inline src/shims/unix/linux/foreign_items.rs:88]: Here, the `if` and `else` have incompatible types. There might be a simple conversion we can wrap the error in.

File: src/shims/unix/macos/foreign_items.rs [hunk 5]
```diff
@@ -180,11 +180,11 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
                 let res = if this.pthread_setname_np(
                     thread,
                     this.read_scalar(name)?,
-                    this.eval_libc("MAXTHREADNAMESIZE").to_target_usize(this)?.try_into().unwrap(),
+                    LibcError("MAXTHREADNAMESIZE").to_target_usize(this)?.try_into().unwrap(),
```
Comment (noahmbright) [inline src/shims/unix/macos/foreign_items.rs:183]: `to_target_usize(this)` is not implemented for `IoError`

File: src/shims/alloc.rs (hunk 1 above)
Comment (RalfJung) [reply src/shims/alloc.rs:112]: Yeah leave this unchanged, it can't use the helper. It doesn't set errno anyway.

File: src/shims/unix/foreign_items.rs (hunk 2 above)
Comment (RalfJung) [reply src/shims/unix/foreign_items.rs:655]: Yes this looks good, except the trailing `?` is missing.

File: src/shims/unix/foreign_items.rs (hunk 3 above)
Comment (RalfJung) [reply src/shims/unix/foreign_items.rs:928]: Leave this unchanged, it doesn't set errno.

File: src/shims/unix/linux/foreign_items.rs (hunk 4 above)
Comment (RalfJung) [reply src/shims/unix/linux/foreign_items.rs:88]: Yeah this doesn't set errno so don't change it.

File: src/shims/unix/linux/sync.rs [hunk 6]
```diff
@@ -153,13 +153,12 @@ pub fn futex<'tcx>(
...
-                    this.eval_libc("ETIMEDOUT"),
+                    LibcError("ETIMEDOUT"),
                 );
             } else {
                 // The futex value doesn't match the expected value, so we return failure
                 // right away without sleeping: -1 and errno set to EAGAIN.
-                let eagain = this.eval_libc("EAGAIN");
-                this.set_last_error(eagain)?;
+                this.set_last_error(LibcError("EAGAIN"))?;
                 this.write_scalar(Scalar::from_target_isize(-1, this), dest)?;
```
Comment (RalfJung) [inline src/shims/unix/linux/sync.rs:162]: This can use `set_last_error_and_return`.

File: src/shims/unix/linux/sync.rs [hunk 7]
```diff
@@ -153,13 +153,12 @@ pub fn futex<'tcx>(
                     Scalar::from_target_isize(0, this), // retval_succ
                     Scalar::from_target_isize(-1, this), // retval_timeout
                     dest.clone(),
-                    this.eval_libc("ETIMEDOUT"),
+                    LibcError("ETIMEDOUT"),
```
Comment (RalfJung) [inline src/shims/unix/linux/sync.rs:156]: This will require `futex_wait` to be changed: the type of `errno_timeout` should become `IoError`.

File: src/shims/unix/macos/foreign_items.rs (hunk 5 above)
Comment (RalfJung) [reply src/shims/unix/macos/foreign_items.rs:183]: Yeah leave these unchanged, they don't set errno.

File: src/shims/unix/solarish/foreign_items.rs
Comment (RalfJung) [inline src/shims/unix/solarish/foreign_items.rs:1]: Also leave these in this file unchanged, again they are not setting errno.

File: src/shims/unix/sync.rs
Comment (RalfJung) [inline src/shims/unix/sync.rs:1]: More things to not change as they are not setting errno.

File: src/shims/alloc.rs (hunk 1 above)
Comment (noahmbright) [reply src/shims/alloc.rs:112]: I can revert this, but I'm a little confused. As far as I can see `this.eval_libc` doesn't set the errno, and `LibcError` is just an enum constant. This change would just be for API consistency, right? Just asking in case this is a change you'd want to make eventually.

Comment (RalfJung) [reply src/shims/alloc.rs:112]: `LibcError` can only be used with `set_last_error`, so only places that set the errno can be converted.

File: src/concurrency/sync.rs [hunk 8]
```diff
@@ -731,9 +731,7 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
                     let futex = this.machine.sync.futexes.get_mut(&addr).unwrap();
                     futex.waiters.retain(|waiter| waiter.thread != thread);
                     // Set errno and write return value.
-                    this.set_last_error(errno_timeout)?;
-                    this.write_scalar(retval_timeout, &dest)?;
-                    interp_ok(())
+                    this.set_last_error_and_return(errno_timeout, &dest)
```
Comment (noahmbright) [inline src/concurrency/sync.rs:734]: Checked the types in futex_wait and made this change - changing the type of errno_timeout broke some other functions downstream: a call to `eval_windows` in the Windows shims passed to futex_wait broke, and in concurrency/thread.rs, a `self.$name.visit_provenance(visit)` isn't implemented for IoErrors. 

I could probably add a `WindowsError` variant to the IoError enum to fix the shims error. I could also fill out more of the Windows error table while I'm at it. Less sure about the architecture of the provenance types, to be honest.

Comment (RalfJung) [reply src/concurrency/sync.rs:734]: Okay let's defer that to a future PR then.

(The right thing would indeed be to add a `WindowsError` variant, and to implement the `VisitProvenance` trait for `IoError`.)

**Comment by RalfJung** [comment]
> marked this pull request as ready for review

Note that this does not trigger a notification, so it is easy to be missed. Please do this instead:
@rustbot ready

File: src/concurrency/sync.rs [hunk 9]
```diff
@@ -731,9 +731,7 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
                     let futex = this.machine.sync.futexes.get_mut(&addr).unwrap();
                     futex.waiters.retain(|waiter| waiter.thread != thread);
                     // Set errno and write return value.
-                    this.set_last_error(errno_timeout)?;
-                    this.write_scalar(retval_timeout, &dest)?;
```
Comment (RalfJung) [inline src/concurrency/sync.rs:735]: This didn't write `-1`, it wrote `retval_timeout`... so your patch changes behavior, and now `retval_timeout` is entirely ignored?

Please revert this part. And generally be more careful about this -- don't touch code you don't understand.

Comment (noahmbright) [reply src/concurrency/sync.rs]: I see, I understand why this is wrong. I'll be more careful. Thank you for the close reviews again.

**Comment by noahmbright** [comment]
Reverted that change in sync.rs and rebased a merge conflict, hence the force push. 
@rustbot ready

**Comment by RalfJung** [comment]
Thanks! Please squash this into a single commit.
@rustbot author

**Comment by noahmbright** [comment]
Done - again thanks for your patience and willingness to help. I was definitely too excited about mindlessly changing every `eval_libc` I could find. This was a good lesson for me. @rustbot ready

**Review by RalfJung — APPROVED** [review]
